Issue #5025 - wrong welcome file handling with dispatcher.include() a…

grgrzybek opened the pull request:
…nd non-default mapping

grgrzybek committed:
Issue #5025 - wrong welcome file handling with dispatcher.include() and non-default mapping

Signed-off-by: Grzegorz Grzybek <gr.grzybek@gmail.com>

joakime commented on jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DefaultServletTest.java:
```diff
@@ -813,6 +815,46 @@ public void testWelcomeMultipleBasesBase() throws Exception
+        HttpTester.Response response;
+
+        // Test included alt default
+        rawResponse = connector.getResponse("GET /context/gateway HTTP/1.0\r\n\r\n");
```
This should be `GET /context/gateway/ HTTP/1.0\r\n\r\n` as you are requesting content belonging to the url-pattern of `/gateway/*` which the request to `/gateway` will not match.

grgrzybek replied:
I checked `org.eclipse.jetty.http.pathmap.PathMappings#getMatch()` and `/p` request is successfully matched for servlet mapped to `/p/*` - it's kept in `org.eclipse.jetty.http.pathmap.PathMappings#_prefixMap`:
```
_prefixMap = {org.eclipse.jetty.util.ArrayTernaryTrie@2636} "{/p=MappedResource[pathSpec=ServletPathSpec["/p/*",pathDepth=2,group=PREFIX_GLOB],resource=default-servlet@2c16bab9==org.ops4j.pax.web.service.jetty.internal.EmbeddedJettyTest$11,jsp=null,order=-1,inst=true,async=true]}"
 LO: int  = 1 (0x1)
 EQ: int  = 2 (0x2)
 HI: int  = 3 (0x3)
 ROW_SIZE: int  = 4 (0x4)
 _tree: char[]  = {char[512]@2893} 
 _key: java.lang.String[]  = {java.lang.String[128]@2894} 
  2 = "/p"
 _value: java.lang.Object[]  = {java.lang.Object[128]@2895} 
  2 = {org.eclipse.jetty.http.pathmap.MappedResource@2643} "MappedResource[pathSpec=ServletPathSpec["/p/*",pathDepth=2,group=PREFIX_GLOB],resource=default-servlet@2c16bab9==org.ops4j.pax.web.service.jetty.internal.EmbeddedJettyTest$11,jsp=null,order=-1,inst=true,async=true]"
   pathSpec: org.eclipse.jetty.http.pathmap.PathSpec  = {org.eclipse.jetty.http.pathmap.ServletPathSpec@2897} "ServletPathSpec["/p/*",pathDepth=2,group=PREFIX_GLOB]"
   resource: java.lang.Object  = {org.eclipse.jetty.servlet.ServletHolder@2656} "default-servlet@2c16bab9==org.ops4j.pax.web.service.jetty.internal.EmbeddedJettyTest$11,jsp=null,order=-1,inst=true,async=true"
```

janbartel replied:
I think you've got 2 differents tests to do here:

1. url is /context/gateway  WITHOUT the trailing slash, which will go to line 377 of ResourceService that does a 302 redirect to url that adds the trailing slash

2. url is /context/gateway/  WITH the trailing slash, which will go to the code at line 402 of ResourceService

janbartel replied:
@grgrzybek I'd still like to see one request with `GET /context/gateway` and one with `GET /context/gateway/`, because the first one will do a redirect before doing the new include path handling: I just want to ensure that we've exercised both codepaths.

joakime commented on jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DefaultServletTest.java:
```diff
@@ -813,6 +815,48 @@ public void testWelcomeMultipleBasesBase() throws Exception
+                req.getRequestDispatcher("/alt/").include(req, resp);
+            }
+        });
+        context.addServlet(gwholder, "/gateway/*");
```
This addServlet should be done before Server.start()

grgrzybek replied:
I don't think it's the case in other tests inside `org.eclipse.jetty.servlet.DefaultServletTest`...

joakime commented on jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DefaultServletTest.java:
```diff
@@ -813,6 +815,48 @@ public void testWelcomeMultipleBasesBase() throws Exception
+        FS.ensureDirExists(altRoot);
+        Path altIndex = altRoot.resolve("index.html");
+
+        ServletHolder defholder = context.addServlet(DefaultServlet.class, "/alt/*");
```
This addServlet should be done before Server.start()

joakime commented:
I'll let others look at this too, but I feel this should be done against branch `jetty-9.4.x` (not `jetty-10.0.x`)

grgrzybek commented:
I'm not sure about the branch. I'm mostly interested in version 9.4.x, but I see that 10.x is the "current" version being developed.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:
```diff
@@ -401,8 +401,13 @@ protected void sendWelcome(HttpContent content, String pathInContext, boolean en
         {
+            String servletPath = included ? (String)request.getAttribute(RequestDispatcher.INCLUDE_SERVLET_PATH)
+                    : request.getServletPath();
+            if (servletPath == null)
```
I think this is a good solution. I would delete the extra "if" at line 406 because that can never happen, unless there's a bug in jetty :)

janbartel reviewed (changes requested):
If you can get rid of the unnecessary "if" statement, and make 2 tests instead of 1, we're golden.

grgrzybek commented:
I've just removed the extra `if` and changed the test, so it uses "strict" mapping for gateway servlet. What's important is not the mapping (and even a request to) for the _gateway_ servlet, but the fact that the servlet calls dispatcher.include()`.

grgrzybek commented:
I've changed the mapping of the _gateway_ servlet from _prefix mapping_ to _strict mapping_ (from `/gateway/*` which also handles `/gateway` requests to plain `/gateway`). According to Servlet spec, `/context/gateway` URL is handled by (mapped to) gateway servlet - there's no redirect at all. Trying `/context/gateway/` request URL simply gives me 404 error.

My point with this test was to check how `/alt/` request works with `dispatcher.include()` - it wasn't meant to test redirect/lack of redirect for the servlet that acts as a gateway.

grgrzybek commented:
When I checked:
```java
context.addServlet(gwholder, "/gateway/*");
```

again, according to my expectations (and Servlet specification), all these requests:
* `/context/gateway`
* `/context/gateway/`
* `/context/gateway/xxx`

are mapped to _gateway_ servlet without redirect and the servlet invokes dispatcher with include for `/alt/` path.
But as I said - mapping and checking the _gateway_ servlet itself was not a point of this test - I wanted to test one line:
```java
req.getRequestDispatcher("/alt/").include(req, resp);
```
:)

janbartel commented:
@grgrzybek oh, I think we're talking about different things. I was wanting to check the behaviour at line 377 of ResourceService (which sends a redirect if the request didn't end in '/').  But, as I've confirmed via a test, the redirect wouldn't work from an include anyway, as include is not permitted to mess with the response headers.

So, to sum up: we're good! Thanks for the PR.

grgrzybek commented:
Thanks @janbartel for review! yes - I went through include + redirect attempts myself in pax-web 8 ;)

grgrzybek commented:
@janbartel May I kindly ask for a backport to Jetty 9.4.x too?

janbartel commented:
@grgrzybek I _knew_ you were going to ask me to do that :).  Done, cherrypicked back to 9.4.x branch just now.

grgrzybek commented:
Ha :) Thanks - much appreciated!

grgrzybek commented:
By the way - I don't have test yet, but when checking Jetty / Tomcat / Undertow, I saw that Jetty sets wrong value to `javax.servlet.include.context_path` attribute when using root context. should be empty string, but is `"/"`. I can think about PR later, but maybe already you have idea if it's wrong? (Undertow and Tomcat set it to `""`).

joakime commented:
@grgrzybek opened issue #5057 about that
